test/common/http/conn_pool_grid_test.cc

```diff
@@ -93,26 +94,43 @@ class ConnectivityGridForTest : public ConnectivityGrid {
 };
 
 namespace {
-class ConnectivityGridTest : public Event::TestUsingSimulatedTime, public testing::Test {
+class ConnectivityGridTestBase : public Event::TestUsingSimulatedTime, public testing::Test {
 public:
-  ConnectivityGridTest()
+  ConnectivityGridTestBase(bool use_alternate_protocols)
       : options_({Http::Protocol::Http11, Http::Protocol::Http2, Http::Protocol::Http3}),
+        alternate_protocols_(simTime()),
         grid_(dispatcher_, random_,
               Upstream::makeTestHost(cluster_, "hostname", "tcp://127.0.0.1:9000", simTime()),
               Upstream::ResourcePriority::Default, socket_options_, transport_socket_options_,
-              state_, simTime(), std::chrono::milliseconds(300), options_),
+              state_, simTime(), maybeCreateAlternateProtocolsCache(use_alternate_protocols),
+              std::chrono::milliseconds(300), options_),
         host_(grid_.host()) {
     grid_.info_ = &info_;
     grid_.encoder_ = &encoder_;
   }
 
+  OptRef<AlternateProtocolsCache> maybeCreateAlternateProtocolsCache(bool use_alternate_protocols) {
+    if (!use_alternate_protocols) {
+      return makeOptRefFromPtr<AlternateProtocolsCache>(nullptr);
+    }
+    return alternate_protocols_;
+  }
+
+  void addHttp3AlternateProtocol() {
+    AlternateProtocolsCache::Origin origin("https", "hostname", 9000);
+    const std::vector<AlternateProtocolsCache::AlternateProtocol> protocols = {
+        {"h3-29", "", origin.port_}};
+    alternate_protocols_.setAlternatives(origin, protocols, simTime().monotonicTime() + Seconds(5));
```

Comment:
nit: let's make this way more than a few seconds to avoid flakes on slow machines?

Reply:
I don't think there's any potential for flakes because this test uses simTime() which is entirely simulated.

Reply:
Ah, perfect.